Add Vibe Mode

aadibajpai opened the pull request:
Implements an automatic loop so you do not need to request lyrics again and again if the song changes, until you stop listening to music or use the $kill command.

This is similar to the way it works in the cli swaglyrics application.

Commit:
implement $vibe tasks loop and commands

pep8speaks commented:
Hello @aadibajpai! Thanks for updating this PR. 

* In the file [`SwagLyricsBot/general_commands.py`](https://github.com/SwagLyrics/SwagLyrics-discord-bot/blob/b39c0375ac052c150b65c4989f7ade05e5ce090d/SwagLyricsBot/general_commands.py):

> [Line 95:66](https://github.com/SwagLyrics/SwagLyrics-discord-bot/blob/b39c0375ac052c150b65c4989f7ade05e5ce090d/SwagLyricsBot/general_commands.py#L95): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace





##### Comment last updated at 2020-08-20 01:32:12 UTC

Commit:
PEP8 fixes

flabbet commented on SwagLyricsBot/general_commands.py:
```diff
@@ -106,10 +103,46 @@ async def send_lyrics():
+
+    @commands.command()
+    async def kill(self, ctx):
+        ctx.send('https://www.youtube.com/watch?v=GF8aaTu2kg0')
```
Ugh, remove this 143 line thing

flabbet reviewed (dismissed):
Well, I know you said it's not going to work with more users, but this approach is a no-go. Using `self` statement makes it almost impossible to make it work for more users. Also, I don't like the idea of calling it vibe mode, Especially in code, what if someone will work on this except us? We would need to explain to him what it is, code should be self-explanatory, this is an unnecessary complication. This is all right to add this as command alias, but not as default one.

Commit:
rework logic acc to discord.py feedback

Commit:
fix conflict

aadibajpai commented:
> Well, I know you said it's not going to work with more users, but this approach is a no-go. Using `self` statement makes it almost impossible to make it work for more users.

fixed

aadibajpai commented:
@thealphadollar, this adds a looping feature to the bot, where you once call the commands and it will keep sending lyrics as activity changes until you stop listening or kill the loop with $kill. The logic is that the $vibe command adds your info to a shared variable and starts a loop if the loop isn't already running. And then the loop every 5s takes care of all the users in the shared variable and $kill removes your info from the shared variable. 

There's a bunch of debug print statements there but I'll get rid of them.

aadibajpai commented:
@dolphingarlic @fadedcoder review again, thanks!

Commit:
works perfectly

looks like there's a problem with the logging part, it doesn't integrate well with this. Once commented it works as intended.

dolphingarlic commented on SwagLyricsBot/general_commands.py:
```diff
@@ -11,9 +11,12 @@
-    def __init__(self, bot, session):
+    def __init__(self, bot, session, current=None):
+        if current is None:
+            current = {}
```
Why not make `current` an empty dict by default?

aadibajpai replied:
because dicts are mutable, so it will persist across inits. https://docs.python-guide.org/writing/gotchas/

dolphingarlic commented on SwagLyricsBot/general_commands.py:
```diff
@@ -59,6 +62,20 @@ def get_spotify_data(user):
         return spotify_activity[0].title, spotify_activity[0].artists
 
+    async def send_lyrics(self, ctx, song, artists):
+
```
Add a docstring

aadibajpai replied:
cool

aadibajpai replied:
done

dolphingarlic commented on SwagLyricsBot/general_commands.py:
```diff
@@ -106,10 +108,61 @@ async def send_lyrics():
             await log.send_webhook()
 
+    @tasks.loop(seconds=5)
+    async def vibe_mode(self):
```
Add a docstring

aadibajpai replied:
cool

aadibajpai replied:
done

dolphingarlic commented on SwagLyricsBot/general_commands.py:
```diff
@@ -106,10 +108,61 @@ async def send_lyrics():
+
+    @commands.command()
+    async def vibe(self, ctx):
+        prev = self.current.copy()
```
Instead of copying the entire dict, you could just check whether the new dict has length equal to 1

dolphingarlic replied:
Also, check whether the user is already in `self.current` if you do this

aadibajpai replied:
alright

aadibajpai replied:
> Also, check whether the user is already in `self.current` if you do this

what do you mean by this

aadibajpai replied:
look at the updated commit and see if that's ok?

dolphingarlic commented on SwagLyricsBot/general_commands.py:
```diff
@@ -106,10 +108,61 @@ async def send_lyrics():
+        # add current user to dict
+        self.current[ctx.author] = {'playing': (None, None),
+                                    'channel': ctx.channel.id}
+        await ctx.send('one vibe mode coming right up.')
```
```suggestion
        await ctx.send('One vibe mode coming right up.')
```

aadibajpai replied:
hmmmm lowercase feels edgy but also inconsistent

dolphingarlic commented on SwagLyricsBot/general_commands.py:
```diff
@@ -106,10 +108,61 @@ async def send_lyrics():
+                await self.send_lyrics(channel, song, artists)
+            except LyricsError as e:
+                print('No activity detected, stopping loop.')
+                await channel.send("No activity detected, killing the vibe.")
```
```suggestion
                await channel.send("No activity detected. Killing the vibe.")
```

aadibajpai replied:
it's not that bad is it?

dolphingarlic commented on SwagLyricsBot/general_commands.py:
```diff
@@ -106,10 +108,61 @@ async def send_lyrics():
+            self.vibe_mode.start()
+
+    @commands.command()
+    async def kill(self, ctx):
```
Maybe rename this to `kill_vibe` or something to prevent ambiguity

aadibajpai replied:
ambiguity with what?

dolphingarlic commented on SwagLyricsBot/lyrics.py:
```diff
@@ -11,7 +11,8 @@ async def fetch(session, url, **kwargs):
     Uses aiohttp to make http GET requests
     """
-    headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.89 Safari/537.36'}
+    headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) '
```
Is the space at the end of this string intentional?

aadibajpai replied:
yeah I just made it go across lines so it's less than 120 chars per line

flabbet commented:
Make logs work (you can't call `add_sub_log` without first calling `add_log`) + clean up console output, and make sure it works _independently_ for every user, also I think this option should be only called on direct messages

flabbet reviewed (changes requested):
Make changes based on the comment above

Commit:
fix logging + address reviews

we don't need to store channel now since DMs only

aadibajpai commented:
> Make logs work (you can't call `add_sub_log` without first calling `add_log`) + clean up console output, and make sure it works _independently_ for every user, also I think this option should be only called on direct messages

check now, also how much console output would you like to keep so it's okay?

flabbet reviewed (changes requested):
Well, it's too much same information for one loop, get rid of it, it should print maybe _ony_  a user name + song name.
And get rid of the slang. Use casual language, but not slang. It's not consistent for the rest of the bot and I will not merge it until you change it. I let you leave $vibe as a command string, but provide aliases like `$loop`, and describe it in `$help` 
![image](https://user-images.githubusercontent.com/25402427/89930651-5e1e0c80-dc0b-11ea-8436-c0caad355386.png)

Oh and fix the conflict

Commit:
Merge branch 'master' into vibe

Commit:
simplify logging, finalize PR

aadibajpai commented:
> Well, it's too much same information for one loop, get rid of it, it should print maybe _ony_ a user name + song name.
> And get rid of the slang. Use casual language, but not slang. It's not consistent for the rest of the bot and I will not merge it until you change it. I let you leave $vibe as a command string, but provide aliases like `$loop`, and describe it in `$help`
> ![image](https://user-images.githubusercontent.com/25402427/89930651-5e1e0c80-dc0b-11ea-8436-c0caad355386.png)
> 
> Oh and fix the conflict

the Spotify thing happens through the logging set up in the preexisting function, so I'm not modifying that. Rest everything has been addressed.

flabbet commented:
Fine, It can stay then

flabbet reviewed (changes requested):
Add the user nickname in this line https://github.com/aadibajpai/SwagLyrics-discord-bot/blob/vibe/SwagLyricsBot/general_commands.py#L49. You should add the log to webhook that someone started vibe and ended it, so we can investigate it easier. Also include the number of people currently in vibe.
